lib/gc/Transforms/Utils/VectorUtils.cpp

```diff
@@ -0,0 +1,526 @@
+      static_cast<uint32_t>((halfValue & 0x7fff) << kF32HalfMantiBitDiff)};
+  const uint32_t exp = shiftedExp & f.u;
+  f.u += kF32HalfExpAdjust; // Adjust the exponent
+
+  // Handle exponent special cases.
+  if (exp == shiftedExp) {
+    // Inf/NaN
+    f.u += kF32HalfExpAdjust;
+  } else if (exp == 0) {
+    // Zero/Denormal?
+    f.u += 1 << kF32MantiBits;
+    f.f -= kF32Magic.f;
+  }
+
+  f.u |= (halfValue & 0x8000) << kF32HalfBitDiff; // Sign bit.
+  return f.f;
+}
+
+// Constructs the 16 bit representation for a bfloat value from a float value.
+// This implementation is adapted from Eigen.
+uint16_t float2bfloat(float floatValue) {
+  if (std::isnan(floatValue))
+    return std::signbit(floatValue) ? 0xFFC0 : 0x7FC0;
+
+  Float32Bits floatBits;
+  floatBits.f = floatValue;
+  uint16_t bfloatBits;
+
+  // Least significant bit of resulting bfloat.
+  uint32_t lsb = (floatBits.u >> kF32BfMantiBitDiff) & 1;
+  uint32_t roundingBias = 0x7fff + lsb;
+  floatBits.u += roundingBias;
+  bfloatBits = static_cast<uint16_t>(floatBits.u >> kF32BfMantiBitDiff);
+  return bfloatBits;
+}
+
+// Converts the 16 bit representation of a bfloat value to a float value. This
+// implementation is adapted from Eigen.
+float bfloat2float(uint16_t bfloatBits) {
+  Float32Bits floatBits;
+  floatBits.u = static_cast<uint32_t>(bfloatBits) << kF32BfMantiBitDiff;
+  return floatBits.f;
+}
+
+std::variant<float, int64_t> numeric_limits_minimum(Type type) {
+  Type t1 = getElementTypeOrSelf(type);
+  if (t1.isF32()) {
+    return -std::numeric_limits<float>::infinity();
+  } else if (t1.isBF16()) {
+    return bfloat2float(float2bfloat(-std::numeric_limits<float>::infinity()));
+  } else if (t1.isF16()) {
+    return (float)half2float(
+        float2half(-std::numeric_limits<float>::infinity()));
+  } else if (t1.isSignedInteger(8)) {
+    return int64_t(-128);
+  } else if (t1.isSignedInteger(32)) {
+    return int64_t(std::numeric_limits<int32_t>::min());
+  } else if (t1.isSignlessInteger(8) or t1.isSignlessInteger(32)) {
+    return int64_t(0);
+  } else {
+    llvm_unreachable("unsupported data type");
+    return (int64_t)0;
+  }
+}
+
+std::variant<float, int64_t> numericLimitsMaximum(Type type) {
+  Type t1 = getElementTypeOrSelf(type);
+  if (t1.isF32()) {
+    return std::numeric_limits<float>::infinity();
+  } else if (t1.isBF16()) {
+    return bfloat2float(float2bfloat(std::numeric_limits<float>::infinity()));
+  } else if (t1.isF16()) {
+    return (float)half2float(
+        float2half(std::numeric_limits<float>::infinity()));
+  } else if (t1.isSignedInteger(8)) {
+    return int64_t(127);
+  } else if (t1.isSignedInteger(32)) {
+    return int64_t(std::numeric_limits<int32_t>::max());
+  } else if (t1.isSignlessInteger(8)) {
+    return int64_t(255);
+  } else if (t1.isSignedInteger(32)) {
+    return int64_t(std::numeric_limits<uint32_t>::max());
+  } else {
+    llvm_unreachable("unsupported data type");
+    return (int64_t)0;
+  }
+}
+
```

Comment:
These are quite general utils and not specific to "vector", I would suggest to move them into a separate file.